### DISCUSSION TITLE
DTOs Wireable and persistence in database

### DISCUSSION THREAD

**Discussion by liepumartins** [livewire/livewire#5194] [1. Help]
The goal is to send as little as necessary data to front-end. From what I have gathered DTOs is the way to do that.

I have some Eloquent models that I would like to partially(only some attributes) present to user and be able to update certain attributes/columns.

If I bind directly to the Eloquent model (making it public in the component) this works. However it sends along all the attributes and meta that is not necessary.
Other option I was tinkering with using `toArray()` and `$visible` / `$hidden` attributes of the model. But then the ability to simply call `$model->save()` is gone.
DTOs are mentioned numerous times, however the documentation is very brief. How to call simple `$model->save()` from a DTO approach? Or have I gotten the idea completely wrong?

````php
<?php

declare(strict_types=1);

namespace App\DTOs;

use App\Models\Cart;
use App\Models\CartItem;
use Livewire\Wireable;

class CartDTO implements Wireable
{
    public function __construct(
        public int $id,
        public string $sum,
        public array $items,
    ){}

    public function toLivewire(): array
    {
        return [
            'id' => $this->id,
            'sum' => $this->sum,
            'items' => $this->items,
        ];
    }

    public static function fromLivewire($value): static
    {
        return new static(
            id: $value['id'],
            sum: $value['sum'],
            items: $value['items'],
        );
    }

    public static function fromModel(Cart $cart): static
    {
        return new static(
            id: $cart->id,
            sum: $cart->sum,            
            items: $cart->items->map(static function (CartItem $item): array {
                return CartItemDTO::fromModel($item)->toLivewire();
            })->toArray(),
        );
    }
}

````

````php
<?php

declare(strict_types=1);

namespace App\DTOs;

use App\Models\CartItem;
use Livewire\Wireable;

class CartItemDTO implements Wireable
{
    public function __construct(
        public int $id,
        public int $cart_id, // not sure if actually necessary
        public string $name,
        public int $quantity,
    ){}

    public function toLivewire(): array
    {
        return [
            'id' => $this->id,
            'cart_id' => $this->cart_id,
            'name' => $this->name,
            'quantity' => $this->quantity,
        ];
    }

    public static function fromLivewire($value): static
    {
        return new static(
            id: $value['id'],
            cart_id: $value['cart_id'],
            name: $value['name'],
            quantity: $value['quantity'],
        );
    }

    public static function fromModel(CartItem $item): static
    {
        return new static(
            id: $item->id,
            cart_id: $item->cart_id,
            name: $item->nomenclature->name,
            quantity: $item->quantity,
        );
    }
}
```` 

And to cry more - if I create DTO (`Livewire\Wireable`) it forces to use array notation in blade :cry: #27

**Comment by joshhanley** [comment]
@liepumartins DTOs a just simple objects that contain data, so you need to write a method that takes the DTO data and saves it back to the model. Sort of like how you have the `fromModel` method to take an Eloquent model and build a DTO from its required data. You need to do the reverse.

You could put this method on the model itself, like `createFromDTO` or `updateFromDTO` or something like that. Or you could even store that on the DTO.

In regards to the array notation in blade when using a wireable DTO, means something isn't right, as for me it works happily using object syntax. I would investigate to see if you can work out why/when it's being converted to an array and see if you can fix that.

**Edit:** I just noticed in your CartDTO, in your fromModel method you are calling `->toLivewire` on the CartItemDTO which will be converting it to an array. Change this to `return CartItemDTO::fromModel($item);` and it should work properly.

Hope this helps!

**Comment by liepumartins** [comment]
Thanks Josh!

I settled for a "hybrid" approach now. 
Cart and items are protected properties, rendered by blade engine.
I created public associative array $quantities. On each update of the array I touch and update corresponding item in the protected property.

```php

    public array $quantities;
    protected Cart $cart;
    protected Collection $items;

    public function booted(): void
    {
        $this->items = $this->cart->items;
        foreach ($this->items as $item) {
            $this->quantities[$item->id] = $item->quantity;
        }
    }

    public function updated()
    {
        $this->validate();
        foreach ($this->quantities as $key => $quantity) {
            $item = $this->items->where('id', $key)->first();
            if ($item->quantity !== intval($quantity)) {
                $item->setQuantityAttribute(intval($quantity));
                $item->save();
            }
        }
        $this->emit('items-changed');
    }
```

This would become more complex and require refactor to DTOs if more attributes were subject to user input.
